### PR TITLE
support the DOS (electronic density of states) fitting

### PR DESCRIPTION
**Pull request by mingzhong15**
here i present the implementation of DeepDOS model [Phys. Rev. B 105, 174109 (2022)] in dpkit-devel branch

generally, i add a new `fitting_type = 'dos'` to allow the DP to fit a vector. this vector can be global (`dos.raw/npy`) or atomic (`atom_dos.raw/npy`), whose length is `numb_dos`.

specifically, 
1. `fit/dos.py`, `loss/dos.py`, `model/dos.py` and `infer/deep_dos.py` is created to allow fitting and inference
2. relevant `*/__init__.py`, `train/trainer.py`, `utils/argcheck.py` is updated
3. in `entrypoint/freeze.py`, `entrypoint/test.py` , relevant functions are added to freeze and test the DeepDOS model
4. a simple user guide is added in the `doc/model/train-fitting-dos.md`
5. an example is provide in the `examples/dos/`, the input script is provided in the `examples/dos/train/input.json`, data sets are provided in the `examples/dos/data/`

### CODE REVIEW

**Comment by njzjz** [comment]
pre-commit.ci autofix

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
1. Please add UTs for your contributions. Thanks!
2. The data in the example are for the demonstrative purposes. Currently they are too large (~600KB) to be included in the source code. Please reduce the size of the data.

File: doc/model/train-fitting-dos.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Fit `DOS`
```
Comment (njzjz) [inline doc/model/train-fitting-dos.md:1]: The full name of DOS should be used in the title, as only some people know it.

File: examples/dos/train/test.sh
Comment (njzjz) [inline examples/dos/train/test.sh:1]: This file is not useful

File: deepmd/fit/dos.py [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+        # args = ()\
+        #        .add('numb_fparam',      int,    default = 0)\
+        #        .add('numb_aparam',      int,    default = 0)\
+        #        .add('neuron',           list,   default = [120,120,120], alias = 'n_neuron')\
+        #        .add('resnet_dt',        bool,   default = True)\
+        #        .add('rcond',            float,  default = 1e-3) \
+        #        .add('seed',             int)               \
+        #        .add("activation_function", str,    default = "tanh")\
+        #        .add("precision",           str, default = "default")\
+        #        .add("trainable",        [list, bool], default = True)
```
Comment (njzjz) [inline deepmd/fit/dos.py:121]: please clean useless comments

File: deepmd/utils/argcheck.py [hunk 3]
```diff
@@ -504,6 +504,50 @@ def fitting_ener():
...
+        Argument("precision", str, optional=True, default="float64", doc=doc_precision),
+        Argument("resnet_dt", bool, optional=True, default=True, doc=doc_resnet_dt),
+        Argument(
+            "trainable", [list, bool], optional=True, default=True, doc=doc_trainable
+        ),
+        Argument("rcond", float, optional=True, default=1e-3, doc=doc_rcond),
+        Argument("seed", [int, None], optional=True, doc=doc_seed),
+        Argument("numb_dos", int, optional=True, default=300, doc=doc_numb_dos),
+        Argument("ener_min", float, optional=True, default=-10.0, doc=doc_ener_min),
+        Argument("ener_max", float, optional=True, default=10.0, doc=doc_ener_max),
```
Comment (njzjz) [inline deepmd/utils/argcheck.py:547]: Is it a good idea to give a default value?

**Review by njzjz — CHANGES_REQUESTED** [review]

**Comment by mingzhong15** [comment]
thank you for the advice, i will revise the code. Also, i will add unit tests

**Review by wanghan-iapcm — APPROVED** [review]

**Comment by njzjz** [comment]
> Patch coverage: **`27.76`**% and project coverage change: **`-1.59`** ⚠️

Please note that only 27.76% of the new codes have been tested.

**Review by njzjz — APPROVED** [review]
@wanghan-iapcm is it ok to merge? Indeed there are still some codes that are not covered.

**Comment by wanghan-iapcm** [comment]
> @wanghan-iapcm is it ok to merge? Indeed there are still some codes that are not covered.

sry i did not notice your comment.

**Pull request merged**